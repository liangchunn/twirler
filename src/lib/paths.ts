import { realpathSync } from 'fs-extra'
import { resolve, join } from 'path'

const appDirectory = realpathSync(process.cwd())
const useTemplate = appDirectory === realpathSync(join(__dirname, '../..'))

function resolveApp(relativePath: string): string {
  return resolve(appDirectory, relativePath)
}

function resolveOwn(relativePath: string): string {
  return resolve(__dirname, '../..', relativePath)
}

export const paths = useTemplate
  ? {
      projectPath: resolveOwn('.'),
      appPackageJson: resolveOwn('package.json'),
      appEntry: resolveOwn('template/src/index.ts'),
      appTsConfig: resolveOwn('template/tsconfig.json'),
      appDevBundle: resolveOwn('template/build/bundle.js'),
      appSrc: resolveOwn('template/src'),
      appProdBundle: resolveOwn('template/build/bundle.prod.js'),
    }
  : {
      projectPath: resolveApp('.'),
      appPackageJson: resolveApp('package.json'),
      appEntry: resolveApp('src/index.ts'),
      appTsConfig: resolveApp('tsconfig.json'),
      appDevBundle: resolveApp('build/bundle.js'),
      appSrc: resolveApp('src'),
      appProdBundle: resolveApp('build/bundle.prod.js'),
    }
