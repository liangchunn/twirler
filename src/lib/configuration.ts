import autoExternal from '@liangchun/rollup-plugin-auto-external'
import { InputOptions, OutputOptions, Plugin, RollupWatchOptions } from 'rollup'
import commonjs from '@rollup/plugin-commonjs'
import del from 'rollup-plugin-delete'
import json from '@rollup/plugin-json'
import resolve from '@rollup/plugin-node-resolve'
import { terser } from 'rollup-plugin-terser'
import typescript from 'rollup-plugin-typescript2'
import { overrideOnWarn } from '../util/overrides'
import { paths } from './paths'

export enum ConfigurationType {
  DEV = 'DEV',
  PROD = 'PROD',
}

export type BaseConfiguration = {
  inputOptions: InputOptions
  outputOptions: OutputOptions
}

export type Configuration =
  | {
      type: ConfigurationType.PROD
      inputOptions: InputOptions
      outputOptions: OutputOptions
    }
  | {
      type: ConfigurationType.DEV
      watchOptions: RollupWatchOptions
    }

export type ConfigurationOptions = {
  dev: boolean
}

export function createBaseConfiguration(
  opts: ConfigurationOptions,
  path: typeof paths
): BaseConfiguration {
  const { dev } = opts

  const inputOptions: InputOptions = {
    input: path.appEntry,
    plugins: [
      del({ targets: 'build/*' }),
      resolve(),
      commonjs(),
      json(),
      autoExternal({
        packagePath: path.appPackageJson,
        resolveOpts: {
          paths: [path.projectPath],
        },
      }),
      typescript({
        tsconfig: path.appTsConfig,
      }),
      !dev && terser(),
    ].filter(Boolean) as Plugin[],
    onwarn: overrideOnWarn,
  }

  const outputOptions: OutputOptions = {
    file: dev ? path.appDevBundle : path.appProdBundle,
    format: 'cjs',
    sourcemap: dev ? 'inline' : true,
  }

  return {
    inputOptions,
    outputOptions,
  }
}

export function createConfiguration(opts: ConfigurationOptions) {
  const { dev } = opts
  const { inputOptions, outputOptions } = createBaseConfiguration(opts, paths)

  return dev
    ? ({
        type: ConfigurationType.DEV,
        watchOptions: {
          ...inputOptions,
          output: outputOptions,
          ...{
            watch: {
              /**
               * TODO: switch to true once rollup can correctly resolve `chokidar`
               * https://github.com/rollup/rollup/issues/2934
               */
              chokidar: false,
            },
          },
        },
      } as const)
    : ({
        type: ConfigurationType.PROD,
        inputOptions,
        outputOptions,
      } as const)
}
