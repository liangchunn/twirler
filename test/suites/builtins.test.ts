import { createBaseConfiguration } from '../../src/lib/configuration'
import { getFixturePaths, bundleFixtures } from '../utils'

describe('builtins', () => {
  it('should not include builtins on dev mode', async () => {
    const fixturePaths = getFixturePaths('builtins')
    const config = createBaseConfiguration({ dev: true }, fixturePaths)
    await bundleFixtures(fixturePaths, config)
  })
  it('should not include builtins on prod mode', async () => {
    const fixturePaths = getFixturePaths('builtins')
    const config = createBaseConfiguration({ dev: false }, fixturePaths)
    await bundleFixtures(fixturePaths, config)
  })
})
