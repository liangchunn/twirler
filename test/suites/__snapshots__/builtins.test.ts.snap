// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builtins should not include builtins on dev mode 1`] = `
"'use strict';

var os = require('os');
var path = require('path');
var fs = require('fs');

function safeFileRead(path, options) {
    try {
        return fs.readFileSync(path, options);
    }
    catch (e) {
        return null;
    }
}

const zshrc = path.join(os.homedir(), '.zshrc');
console.log(safeFileRead(zshrc, { encoding: 'utf-8' }));
"
`;

exports[`builtins should not include builtins on prod mode 1`] = `
"\\"use strict\\";var r=require(\\"os\\"),e=require(\\"path\\"),n=require(\\"fs\\");const t=e.join(r.homedir(),\\".zshrc\\");console.log(function(r,e){try{return n.readFileSync(r,e)}catch(r){return null}}(t,{encoding:\\"utf-8\\"}));
"
`;
